blockprod/src/rpc.rs

```diff
@@ -22,54 +22,59 @@ use common::{
 
 use serialization::{hex::HexDecode, hex::HexEncode};
 
-use crate::BlockProductionError;
+use crate::{detail::JobKey, BlockProductionError};
 use subsystem::subsystem::CallError;
 
 #[rpc::rpc(server, namespace = "blockprod")]
 trait BlockProductionRpc {
     /// Stop block production
-    #[method(name = "stop")]
-    async fn stop(&self) -> rpc::Result<()>;
+    #[method(name = "stop_all")]
+    async fn stop_all(&self) -> rpc::Result<()>;
 
-    /// Start block production on the next chance (when new tip is available)
-    #[method(name = "start")]
-    async fn start(&self) -> rpc::Result<()>;
+    // Stop a specific job
+    #[method(name = "stop_job")]
+    async fn stop_job(&self, job_id: JobKey) -> rpc::Result<()>;
 
     /// Generate a block with the supplied transactions to the specified reward destination
+    /// If transactions are None, the block will be generated with available transactions in the mempool
     #[method(name = "generate_block")]
     async fn generate_block(
         &self,
         reward_destination_hex: String,
-        transactions_hex: Vec<String>,
+        transactions_hex: Option<Vec<String>>,
     ) -> rpc::Result<String>;
 }
 
 #[async_trait::async_trait]
 impl BlockProductionRpcServer for super::BlockProductionHandle {
-    async fn stop(&self) -> rpc::Result<()> {
-        handle_error(self.call(|this| this.stop()).await)
+    async fn stop_all(&self) -> rpc::Result<()> {
+        handle_error(self.call_mut(|this| this.stop_all()).await)
     }
 
-    async fn start(&self) -> rpc::Result<()> {
-        handle_error(self.call(|this| this.start()).await)
+    async fn stop_job(&self, job_id: JobKey) -> rpc::Result<()> {
+        handle_error(self.call_mut(move |this| this.stop_job(job_id)).await)
     }
 
     async fn generate_block(
         &self,
         reward_destination_hex: String,
-        transactions_hex: Vec<String>,
+        transactions_hex: Option<Vec<String>>,
     ) -> rpc::Result<String> {
```

Comment:
Currently `generate_block()` works as "call function -> block is returned", however with the introduction to `JobKey`, we should probably refactor this API - possibly something like splitting this into multiple functions as there's currently no way for the caller to know what its JobKey is so that it can call `stop_job()`. I was thinking something like:

1. `generate_block()` as is for PoW calls
2. `enqueue_block_generation()` so that PoS can get a JobKey of its pending job
3. `poll_block_generation()` so that PoS can check/get results of its pending job

Reply:
This is something to be fixed. I did the bare minimum. Please feel free to improve this design.

Reply:
Since it will change for PoS, this can be left as it is now.